## Data Analysis I

Lastly, we will answer the following questions using pandas methods: 

* What are the most popular colors by season?
* What is the most popular clothing item by season?
* What is the effect of promo-codes on the dollar-amount of purchases?
* When do users leave a review?
* Do states vary in the items they purchase in the Fall?

Utilize the documentation provided in each code-block. When you are done with this section of the project, validate that your output matches the screenshot provided in the `docs/part3.md` file and answer the questions located underneath `Data Analysis II` in your own words.

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from scipy import stats

In [2]:
# TODO: load `data/processed/shopping_cleaned.csv` as a pandas dataframe
# df=pd.read_csv('C:/Users/deema/Desktop/Lab4/shopping-behavior/code/shopping_cleaned.csv')

df=pd.read_csv('C:/Users/deema/Desktop/Lab4/shopping-behavior/data/processed/shopping_cleaned.csv')

print(df.shape)


(3158, 13)


In [3]:
# TODO: Create a dataframe pivot table where "Season" is your column, "Color" is your index, and "Location" is your value (this value is arbitrary)
# use the len function as the "aggfunc" parameter.
# Save this pivot table to a new variable
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html

Nu_1_pivot_table = df.pivot_table(index='Color', columns='Season', values='Location', aggfunc=len, fill_value=0)


In [4]:
# TODO: Display this pivot table

print(Nu_1_pivot_table)


Season                Fall  Spring  Summer  Winter
Color                                             
Aubergine               83       2      13     111
Baby blue               16      69      84      21
Black                   84       5       9     124
Brick red               88       7      12     110
Brown                   97       5       3      89
Burnt orange            93       4       5     112
Fuchsia                  3      45      78      18
Lavender                13      55     102       6
Lemon yellow             9      47      98      13
Mauve                   86       5       7      95
Muted mustard yellow    84       8      13      87
Pale peach               9      52      81      15
Periwinkle               9      54      82       8
Ruby Red                90       3       8     107
Terra cotta             91       4      15     101
Turquoise                7      49      74      16
White                   10      63      81      11


In [5]:
# TODO: Select the "Fall" column from this pivot table and display the sort order in descending order
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html

sorted_column = Nu_1_pivot_table['Fall'].sort_values(ascending=False)

print(sorted_column)

Color
Brown                   97
Burnt orange            93
Terra cotta             91
Ruby Red                90
Brick red               88
Mauve                   86
Black                   84
Muted mustard yellow    84
Aubergine               83
Baby blue               16
Lavender                13
White                   10
Periwinkle               9
Pale peach               9
Lemon yellow             9
Turquoise                7
Fuchsia                  3
Name: Fall, dtype: int64


In [6]:
# TODO: Select the "Winter" column from this pivot table and display the sort order in descending order.

sorted_column = Nu_1_pivot_table['Winter'].sort_values(ascending=False)
print(sorted_column)

Color
Black                   124
Burnt orange            112
Aubergine               111
Brick red               110
Ruby Red                107
Terra cotta             101
Mauve                    95
Brown                    89
Muted mustard yellow     87
Baby blue                21
Fuchsia                  18
Turquoise                16
Pale peach               15
Lemon yellow             13
White                    11
Periwinkle                8
Lavender                  6
Name: Winter, dtype: int64


In [7]:
# TODO: Select the "Spring" column from this pivot table and display the sort order in descending order

sorted_column = Nu_1_pivot_table['Spring'].sort_values(ascending=False)
print(sorted_column)

Color
Baby blue               69
White                   63
Lavender                55
Periwinkle              54
Pale peach              52
Turquoise               49
Lemon yellow            47
Fuchsia                 45
Muted mustard yellow     8
Brick red                7
Mauve                    5
Brown                    5
Black                    5
Burnt orange             4
Terra cotta              4
Ruby Red                 3
Aubergine                2
Name: Spring, dtype: int64


In [8]:
# TODO: Select the "Summer" column from this pivot table and display the sort order in descending order

sorted_column = Nu_1_pivot_table['Summer'].sort_values(ascending=False)
print(sorted_column)

Color
Lavender                102
Lemon yellow             98
Baby blue                84
Periwinkle               82
Pale peach               81
White                    81
Fuchsia                  78
Turquoise                74
Terra cotta              15
Muted mustard yellow     13
Aubergine                13
Brick red                12
Black                     9
Ruby Red                  8
Mauve                     7
Burnt orange              5
Brown                     3
Name: Summer, dtype: int64


In [9]:
# TODO: Using the "idmax" method, get the index with the maximum value for each column
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html

Nu_1_pivot_table.idxmax()



Season
Fall          Brown
Spring    Baby blue
Summer     Lavender
Winter        Black
dtype: object

In [10]:
# TODO: Create a dataframe pivot table where "Season" is your column, "Item Purchased" is your index, and "Location" is your value (this value is arbitrary)
# use the len function as the "aggfunc" parameter.
# Save this pivot table to a new variable and display it

Second_pivot_table = df.pivot_table(index='Item Purchased', columns='Season', values='Location', aggfunc=len, fill_value=0)

In [11]:
# TODO: Using the "idmax" method, get the index with the maximum value for each column on the pivot table containing clothing items & season
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html

Second_pivot_table.idxmax()

Season
Fall           Backpack
Spring    Running Shoes
Summer           Shorts
Winter         Leggings
dtype: object

In [12]:
# TODO: Create two data-frames for promo-code users and non-promo-code users using boolean indexing
# Save these two data frames into separate variables 
# Documentation: https://pandas.pydata.org/docs/user_guide/indexing.html#boolean-indexing
# Hint: Users that have used a promo code have the value "Yes" in the column "Promo Code Used"

#The following code is trying to do:
#Make 2 New DataFrame for the original data
#Using those 2 data frames create 2 new sets of data  frames that are reading the "Promo Code Used" field
#Evaluating the column to determine "TRUE" or "FALSE"
#That is "Yes" or "No" indicating rather or not the promo code was used.


#Make a 2 copies of the data frame
df_backcopy_1 = df.copy()
df_backcopy_2 = df.copy()


promo_code_users_df = df_backcopy_1[df_backcopy_1['Promo Code Used'] == 'Yes']
non_promo_code_users_df = df_backcopy_2[df_backcopy_2['Promo Code Used'] == 'No']



In [13]:
# TODO: Print the shape of the promo code users dataframe. This should be (1373, 13)

print(promo_code_users_df.shape)



(1373, 13)


In [14]:
# TODO: Print the average purchase size of promo code users dataframe. This should be 50.06747998236351

# Filter the DataFrame to include only the rows where a promo code was used
promo_code_users = df_backcopy_1[df_backcopy_1['Promo Code Used'] == 'Yes']

# Calculate and print the average purchase size
average_purchase_size = promo_code_users['Purchase Amount (USD)'].mean()


print(average_purchase_size)



50.06747998236351


In [15]:
# TODO: Print the shape of the non promo code users dataframe. This should be (1785, 13)

print(non_promo_code_users_df.shape)


(1785, 13)


In [16]:
# TODO: Print the average purchase size of non promo code users dataframe. This should be 30.155750765104113


# Filter the DataFrame to include only the rows where a promo code was used
non_promo_code_users = df_backcopy_2[df_backcopy_2['Promo Code Used'] == 'No']

# Calculate and print the average purchase size
average_purchase_size = non_promo_code_users['Purchase Amount (USD)'].mean()
print(average_purchase_size)


30.155750765104113


Formulate a null hypothesis regarding the dollar amount spent between promo-code users & non-promo-code users

One null hypothesis could be that the total dollars that those customer with a promo code (promo code user) would spend more money on average versus the customer that did not have a promo code (non-promo code user).

Formulate an alternative hypothesis regarding the dollar amount spent between promo-code users & non-promo-code users

Alternately, if the customer did have a promo code, it may only be applicable to a certain amount of money that the customer will have to spend in order to even have the ability to use the promo code. e.g a promo code will be applicable for all orders OVER 200.00 Dollars. Therefore, if the customer is unable to afford to purchase that amount of merchandise totalling over 200.00 Dollars, they will be automatically restricted from using the promo code.

In [17]:
# TODO: Using the "ttest_ind" method from "scipy.stats", run a T-Test between the dollar amount spent of promo-code users and non promo-code users
# The t-statistic you should observe is 93.2464714350831
# Documentation: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html

#ttest for 1.) dollar amount spent of promo-code users and 2.) non promo-code users

# Assuming this is your data
#promo_code_users = [50.06747998236351]  # fill this with the dollar amount spent of promo-code users
#non_promo_users = [30.155750765104113]  # fill this with the dollar amount spent of non promo-code users

import scipy
from scipy import stats

# Assuming the following lists contain the dollar amount spent by each group
promo_code_users = [50.06747998236351] 
non_promo_code_users = [30.155750765104113] 

# Perform the T-Test
t_statistic, p_value = stats.ttest_ind(promo_code_users, non_promo_code_users)

print(f'T-statistic: {t_statistic}')
print(f'P-value: {p_value}')

T-statistic: nan
P-value: nan


c:\Users\deema\anaconda3\envs\phase1\lib\site-packages\scipy\stats\_stats_py.py:7030: RuntimeWarning: invalid value encountered in scalar divide
  svar = ((n1 - 1) * v1 + (n2 - 1) * v2) / df


In [18]:
# TODO: Group your dataframe according to the "Review Rating" column. This should produce two groups for the values "Missing" & "Present"
# Save this grouped dataframe into a new variable named "g_review"
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html

df=pd.read_csv('C:/Users/deema/Desktop/Lab4/shopping-behavior/code/shopping_cleaned.csv')
g_review = df.groupby('Review Rating')
print(df['Review Rating'].head()) 

0        4.0
1    Missing
2    Missing
3        5.0
4        2.0
Name: Review Rating, dtype: object


In [19]:
# TODO: Print out the keys of this dataframe for validation. This should output "dict_keys(['Missing', 'Present'])"
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.keys.html

print(g_review.groups.keys())


dict_keys(['1.0', '2.0', '3.0', '4.0', '5.0', 'Missing'])


In [20]:
# TODO: calculate the average of the "Purchase Amount (USD)", "Age", and "Previous Purchases" columns in our grouped dataframe
# output the values for display
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.mean.html
# Documentation: https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html

#The average values from the grouped by range for 'Purchase Amount (USD)', 'Age', 'Previous Purchases

average_values = g_review[['Purchase Amount (USD)', 'Age', 'Previous Purchases']].mean()

# Print the average values
print(average_values)

               Purchase Amount (USD)        Age  Previous Purchases
Review Rating                                                      
1.0                        34.145592  32.945946            4.270270
2.0                        38.813979  34.042169            6.415663
3.0                        32.777565  38.000000            0.000000
4.0                        38.699208  35.147627            6.952548
5.0                        38.865241  33.446575            6.945205
Missing                    38.923623  34.080198            5.344059


In [21]:
# TODO: Create a new column called "Prev Purchase Group" on your original dataframe which seperates your dataframe into 
# two groups of shoppers based on their "Previous Purchases" column using the "pd.cut" method. 
# Ensure that you are only creating 2 "bins" and label these respective bins as ["Low", "High"]
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.cut.html

df_copy=pd.read_csv('C:/Users/deema/Desktop/Lab4/shopping-behavior/code/shopping_cleaned.csv')

#df=pd.read_csv('C:/Users/deema/Desktop/Lab4/shopping-behavior/code/shopping_cleaned.csv')

df_copy['Prev Purchase Group'] = pd.cut(df_copy['Previous Purchases'], bins=2, labels=['Low', 'High'])
print(df_copy['Prev Purchase Group'].head(10))

0    Low
1    Low
2    Low
3    Low
4    Low
5    Low
6    Low
7    Low
8    Low
9    Low
Name: Prev Purchase Group, dtype: category
Categories (2, object): ['Low' < 'High']


In [29]:


# TODO: Using the "value_counts" function, count how many Missing & Present values are in the "Low" group
# Display this value for analysis 
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html
# Hint: You will have to use Boolean Indexing

#Show the 'Low' group
low_group = df_copy[df_copy['Prev Purchase Group'] == 'Low']

#Count the 'Missing' and 'Present' values
value_counts = low_group['Review Rating'].value_counts()

#Print the counts
print(value_counts)

Review Rating
Missing    1686
4.0         412
5.0         263
2.0         132
1.0          32
3.0           1
Name: count, dtype: int64


In [30]:
# TODO: Using the "value_counts" function, count how many Missing & Present values are in the "High" group
# Display this value for analysis 
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html
# Hint: You will have to use Boolean Indexing

#Show the 'High' group
high_group = df_copy[df_copy['Prev Purchase Group'] == 'High']

#Count the 'Missing' and 'Present' values
value_counts = high_group['Review Rating'].value_counts()

# Print the counts
print(value_counts)

Review Rating
Missing    334
4.0        157
5.0        102
2.0         34
1.0          5
Name: count, dtype: int64


## Data Analysis II

In the next section, answer the primary analytical questions in your own words.

### Q1

What are the top three colors for Fall & Winter?

Top 3 colors for Fall and Winter are: Fall ( there were some ties so I listed them): 
Fall
Brown                   97
Burnt Orange            93
Terra Cotta             91

Winter
Black                   124
Burnt Orange            112
Aubergine               111



### Q2

What are the top three colors for Spring & Summer?

Top 3 colors for Spring and Summer are: 
Spring: 
Baby blue               69
White                   63
Lavender                55

Summer:
Lavender                102
Lemon yellow             98
Baby blue                84



### Q3

What is the most popular clothing item by season?

Season
Fall           Backpack      
Spring    Running Shoes
Summer           Shorts  
Winter         Leggings

### Q4

Observing the p-value that we got when running our t-test between promo-code and non-promo-code users, what can we conclude regarding our null hypothesis?

The Verdict is still out! 

### Q5

Observe the `value_counts` output for the "Review Rating" column for both your "High" and "Low" groups. Proportionally speaking, which group is more likely to leave a review? Why might this be happening from the "human"-perspective? Rationalizations are ok at this point, even if they aren't backed up by data.

I believe that the group most likely to leave a review at all is: Low group. Our dollars count for something, no matter how little we spend. I think that the economy is driving a lot of these factors. Everything is more expensive like 2-3 times as great as it was 2-3 years ago. And even when you are paying less, you are still expecting to have Great Value, Great Product Satisfaction, Great Reliability from your purchase. I would be interested to learn what the ratings were prior to the pandemic and even prior to 2000.